Support Linux on z as a Swift platform

bryanpkc opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->
This pull request adds support for Linux on IBM z Systems to Swift. It depends on pull requests in the other repos:

apple/swift-clang#13
apple/swift-llvm#8
apple/swift-lldb#27

The first commit in this pull request adds a new platform condition check named `endian()`, for which there are two possible values (`little` and `big`), to allow conditional compilation of endianness-specific Swift code. The second commit constitutes the bulk of infrastructure changes, including new test cases, to add support for the new target. The remaining commits deal with endianness issues in various parts of the compiler and Swift runtime. In particular, handling of enums is not yet completely fixed for big-endian systems, and we are researching a full solution (to be contributed in a future pull request). Nevertheless, the port gives us a usable compiler and REPL. Validation test results show a 0.2% failure rate:

```
  Expected Passes    : 7308
  Expected Failures  : 71
  Unsupported Tests  : 797
  Unexpected Failures: 19
```

No regression is observed on x86_64.
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

gribozavr commented:
> The first commit in this pull request adds a new platform condition check named endian(), for which there are two possible values (little and big), to allow conditional compilation of endianness-specific Swift code.

This change needs to go through the Swift evolution process.  We can unblock the port by adding it as `_endian` for now.

gribozavr commented on stdlib/private/SwiftPrivateLibcExtras/SwiftPrivateLibcExtras.swift:
```diff
@@ -81,6 +81,12 @@ public struct _stdlib_fd_set {
 
+#if arch(arm) || arch(i386)
+public typealias _stdlib_fd_set = __stdlib_fd_set<UInt32>
+#else
```
Please list 64-bit platforms explicitly.  This ensures that this code breaks when we have a new port, and it is added to the appropriate list.

gribozavr commented on stdlib/private/SwiftPrivateLibcExtras/SwiftPrivateLibcExtras.swift:
```diff
@@ -37,36 +37,36 @@ public var _stdlib_FD_SETSIZE: CInt {
 
-public struct _stdlib_fd_set {
-  var _data: [UInt32]
+public struct __stdlib_fd_set<T: UnsignedInteger> {
```
Why was this change required?

gribozavr replied:
Also, the standard library style is to put spaces on both sides of the colon in generic constraints (`<T : UnsignedInteger>`).

bryanpkc replied:
The original implementation aliases an array of C long int's with an array of Swift UInt32's, regardless of architecture word size, which happened to work fine for little-endian systems but not for big-endian ones.

gribozavr replied:
I see now why it is required -- we are passing the buffer directly to the select syscall.  Existing code would work everywhere except on 64-bit big endian platforms, right?

Instead of making it generic, which feels unnecessary to me -- and allowing creating invalid combinations of types for a platform, could we conditionally define a typealias for the underlying type (`UInt32` on 32-bit, `UInt64` on 64-bit)?  You could probably even just use a plain `UInt` without any conditionals.

gribozavr commented on stdlib/public/SDK/CoreGraphics/CGFloat.swift.gyb:
```diff
@@ -28,7 +28,7 @@ public struct CGFloat {
   /// 32-bit architectures and Double on 64-bit architectures.
   public typealias NativeType = Float
-#elseif arch(x86_64) || arch(arm64)
+#elseif arch(x86_64) || arch(arm64) || arch(s390x)
```
This overlay is specific to Apple platforms, so this change shouldn't be needed.

gribozavr commented on stdlib/public/core/Builtin.swift:
```diff
@@ -395,6 +395,19 @@ internal var _objectPointerLowSpareBitShift: UInt {
+    @inline(__always) get { return 0 }
+}
+internal var _objCTaggedPointerBits: UInt {
+    @inline(__always) get { return 0 }
```
Indentation is 2 spaces.

gribozavr commented on stdlib/public/core/Character.swift:
```diff
@@ -115,19 +116,40 @@ public struct Character :
       _representation = .large(nativeString._core.nativeBuffer!._storage)
     }
+#else
+    // On big-endian systems (e.g. s390x), disable the optimization for now.
```
This is a major performance hit for the common case.  Please leave a `FIXME(performance)` marker.

gribozavr commented on stdlib/public/core/FixedPoint.swift.gyb:
```diff
@@ -322,7 +322,7 @@ public struct ${Self}
 #if arch(i386) || arch(x86_64) || arch(arm) || arch(arm64) || arch(powerpc64le)
     return self
-#elseif arch(powerpc64)
+#elseif arch(powerpc64) || arch(s390x)
```
You can actually change all of these lists to `endian(little)` / `endian(big)`.

gribozavr commented on stdlib/public/core/StringUTF8.swift:
```diff
@@ -236,7 +245,13 @@ extension String {
+      let numBits = (sizeof(_StringCore._UTF8Chunk.self) &- 1) &* 8
+      let hiByte = position._buffer >> numericCast(numBits)
+      let result = UTF8.CodeUnit(truncatingBitPattern: hiByte & 0xFF)
+#endif
```
Are these differences in string buffer handling caused by different layout of bits in the 64-bit `_UTF8Chunk`?  Could we instead change the code that produces these values to produce a consistent layout on all platforms?

bryanpkc replied:
Yes, on big-endian systems the bytes in a `_UTF8Chunk` are laid out in the opposite order. I will try to improve this code.

gribozavr replied:
That would be great.  We are expecting to refactor the String internals soon, and that will certainly break big-endian systems if any code using small UTF-8 buffers needs to be different.

gribozavr commented on stdlib/public/runtime/HeapObject.cpp:
```diff
@@ -229,13 +229,15 @@ static Lazy<MetadataCache<BoxCacheEntry>> Boxes;
 extern "C"
-BoxPair::Return SWIFT_RT_ENTRY_IMPL(swift_allocBox)(const Metadata *type) {
+BoxPair::Return SWIFT_RT_ENTRY_IMPL(swift_allocBox)(const Metadata *type)
+    SWIFT_CC(swift) {
```
I thought the dominant style is to put `SWIFT_CC(swift)` on a separate line before the declaration -- doesn't that work in this case?

gribozavr commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -464,6 +464,8 @@ - (BOOL)isNSValue__      { return NO; }
 #elif defined(__arm64__)
 static uintptr_t const objectPointerIsObjCBit = 0x4000000000000000ULL;
+#elif defined(__s390x__)
+static uintptr_t const objectPointerIsObjCBit = 0x4000000000000000ULL;
```
This change shouldn't be needed, since this code is behind `#if SWIFT_OBJC_INTEROP`.

gribozavr commented on test/Parse/ConditionalCompilation/endianness.swift:
```diff
@@ -0,0 +1,19 @@
+// CHECK:      [[X:%[0-9]+]] = integer_literal $Builtin.Int1, -1
+// CHECK-NEXT: {{%[0-9]+}} = struct $Bool ([[X]] : $Builtin.Int1)
+
+#if arch(arm) || arch(arm64) || arch(i386) || arch(x86_64) || arch(powerpc64le)
```
I'd recommend adding the endianness checks to each file in `test/Parse/ConditionalCompilation` instead (just after the `_runtime()` clause).  This ensures that the tests run always, and not just when targeting that platform.

gribozavr commented on stdlib/public/runtime/Enum.cpp:
```diff
@@ -154,12 +154,16 @@ swift::swift_getEnumCaseSinglePayload(const OpaqueValue *value,
                                        1 /*payload case*/);
 
+#if defined(__BIG_ENDIAN__)
+    small_memcpy(reinterpret_cast<uint8_t*>(&extraTagBits) + 4 - numBytes,
```
Where does the magic number 4 come from?

bryanpkc replied:
It comes from the fact that the extra tag can only be 1 byte, 2 bytes or 4 bytes in size (according to the `getNumTagBytes` function). I agree that this limit can be stated more clearly in the code.

gribozavr commented:
CC @rjmccall @slavapestov for IRGen and runtime changes.

The library parts and test changes LGTM.

gribozavr commented:
@bryanpkc This is fantastic!  I'm just curious -- what are the failing tests?

gribozavr commented:
Let's run CI to verify that there are no regressions on existing platforms.

@swift-ci Please test

gribozavr commented:
@bryanpkc Looks like there are test failures on OS X, please take a look.

bryanpkc commented:
@gribozavr Thanks for your review. I have made changes according to your suggestions and fixed the test failure on OS X.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@bryanpkc OS X failed again.

lattner commented:
@gribozavr thank you for the extensive review, I really appreciate it.   @bryanpkc please do submit a proposal for adding the endian #if check.  I expect it to be mostly a bike shedding exercise, but it is an important one to do and get past.

rintaro commented:
@bryanpkc 
FYI, http://thread.gmane.org/gmane.comp.lang.swift.evolution/12267

bryanpkc commented:
@gribozavr I think I have fixed the OS X test case for real this time (sorry I cannot run that test myself). I have also simplified the StringUTF8 fix as you suggested. It also helps us pass a few more test cases.

@lattner, @rintaro: Thanks for the link to the draft proposal. It seems that the bitwidth() part of the proposal is more complicated and warrants more discussion, but I can submit the endian() part of it, if @sadun doesn't mind.

rintaro commented on stdlib/public/core/StringUTF8.swift:
```diff
@@ -48,8 +48,10 @@ extension _StringCore {
 
-      return (i + utf16Count, result)
+      // Convert the _UTF8Chunk into host endianness.
+      return (i + utf16Count, _UTF8Chunk(littleEndian: result))
```
I think, we have another endianness problem in `_StringCore._nthContiguous`.
Sorry, I cannot test myself.

bryanpkc replied:
@rintaro You are right about `_nthContiguous()` being little-endian only but that is only part of a larger problem. There does not seem to be any support for UTF-16BE in Swift, and supporting it may require considerable changes to the _StringCore code. I am pretty sure that conversion from UTF-8 to UTF-16 is currently broken in this port, but we hope to tackle that in a future pull request (soonish).

gribozavr commented:
@bryanpkc Thank you!

@swift-ci Please test

gribozavr commented:
@swift-ci Please test OS X platform

gribozavr commented:
@swift-ci Please test OS X platform

gribozavr commented:
@swift-ci Please test

gribozavr commented:
The CI failed for unrelated reasons.

Ping @rjmccall @slavapestov for IRGen and runtime changes.

bryanpkc commented:
I have rebased the commits on to the tip of master, and added a couple more fixes for test failures.

gribozavr commented:
@slavapestov has reviewed the changes.

gribozavr commented:
@swift-ci Please test and merge

bryanpkc commented:
@gribozavr Thanks for testing. The TestNSJSONSerialization failure does not seem related to my commits.

gribozavr commented:
Yes, it isn't.

bryanpkc commented:
Thanks for merging.
